alacritty_terminal/src/tty/windows/mod.rs

```diff
@@ -306,3 +323,17 @@ impl EventedPty for Pty {
         }
     }
 }
+
+impl Drop for Pty {
+    fn drop(&mut self) {
+        if let PtyHandle::Conpty(c) = &self.handle {
+            // The conpty close call will block until the conout pipe is drained.
+            //
+            // EventedAnonRead has an internal buffer which is continuously draining the conout
+            // pipe into, so as long as this `.close()` call is run before self.conout is dropped
+            // it won't lead to a deadlock. (It will block just for a moment while EventedAnonRead
+            // does its internal workings.)
+            c.close();
+        }
+    }
+}
```

Comment:
Okay, so having to call `close` here with runtime-enforced crash in the `Drop` for the `ConPTY` is a bit problematic. We kinda use Rust to make sure things are enforced at compile-time as well as possible and this pushes a lot of the checks to runtime which certainly isn't ideal or intuitive.

I'd assume the problem arose because it's not possible to manually specify the drop order in the `Drop` impl for `Pty`, right? So you weren't able to do what you're currently doing in `close` in the `Drop` impl for `ConPTY`?

Rust actually has a mechanism to change that, which would allow us to have a well defined drop order while still relying on the standard `Drop` trait instead of custom `close` functions. This is documented in https://github.com/rust-lang/rfcs/blob/master/text/1857-stabilize-drop-order.md.

The short version is that variables that are dropped in the order they are defined, like this:

```rust
struct X {
    a: String, // Dropped First
    b: String, // Dropped Second
    c: String, // Dropped Third
}
```

So I think we could save ourselves a lot of runtime logic by making use of this well defined behavior. We probably wouldn't need any of the atomic operations, or the strange `close` method if done like this.

To prevent people from destroying things, we could just add a comment to the `PTY`'s handler field that explains this.

Reply:
I'm happy to rely on drop order!

To do that, I'll have to change `PtyHandle` to store an actual `Conpty` rather than `ConptyHandle` (which is `Arc<Conpty>`), so that dropping the `Pty` actually drops the `Conpty` rather than just decreasing the reference count.

This will lead to a little bit of fallout:
- Because doing the above will mean we won't have `Conpty` in an `Arc`, I'll have to add a `ConptyResizeHandle` type which implements `OnResize`. Dropping a `ConptyResizeHandle` won't close the pseudoconsole.
- Change the `Pty::resize_handle` function - because it'll need to return `WinptyHandle` for WinPTY, and `ConptyResizeHandle` for ConPTY. Maybe the best option is for it to return `Box<impl OnResize>`?

It's a mess indeed! I can do some follow-up PRs to solve this tangle of WinPTY / ConPTY code.

Sound good to you?

Reply:
> To do that, I'll have to change PtyHandle to store an actual Conpty rather than ConptyHandle (which is Arc<Conpty>), so that dropping the Pty actually drops the Conpty rather than just decreasing the reference count.

That's a good point, I haven't thought about this. It also *again* makes things a lot more complicated of course.

We won't be able to have `resize_handle` just be `self.handle.clone()` anymore, since the handle itself cannot be shared without any issues so we'd probably have to create a resize handle trait or something like that which the PTY backends would have to implement.

We'd then also have to make sure that when resizing, the ConPTY hasn't been closed yet, adding yet more friction and caveats to the ConPTY API.

From my understanding, what we would want is more similar to just packing `Pty` in an `Arc` instead of the `ConPTY` handle and making sure we keep the entire `Pty` around until it can be destroyed. However we cannot do that since there are a lot of other fields on the `Pty` which we do not want to reference count.

Resizing is also the only thing the handle is ever used for. So a design which does not implement drop for the PtyHandle, but then stores the handle on the `Pty` just to ensure the `ConPTY` is dropped at the right time, is probably a bit strange and confusing.

I'm wondering if the 'resize from anywhere' structure of the resize handle is a good fit at all for ensuring these drop order invariants. In all of the current approaches we had some kind of `closed` atomic, or other various bits that ensured that the resize handle doesn't explode after the PTY was closed.

We already have a channel which we use in `event.rs` to write info to the pty, I wonder if we could use that same channel to send resize messages through. Then the `event_loop.rs` would be the sole owner of the `pty` and we could just read new messages from the channel and directly call `pty.resize(xxx)`. That way we do not need an `Arc` anymore and all our problems would be gone.

However I honestly do not want to keep suggesting changes with this PR just to see how they do not work out once implemented. I think the best way forward is to go for the simplest and smallest fix possible to resolve this issue. If we want to actually fix things, we need to refactor code and that should come after fixing this bug.

That probably means just manually calling `drop(pty)` before the program exits, adding a comment above that why we need it and then we automatically get the correct behavior since our drop order is correct. The resize_handle staying around after that doesn't matter at all, since it will not receive any more events because everything is shut down already anyways, so all it does is delay the close call on the ConPTY which doesn't hurt anything. Just to make sure, I'd still reorder the fields in `ConPTY` just to make sure that if the `resize_handle` is dropped first, everything still works fine. And then we should be golden.

Reply:
> However I honestly do not want to keep suggesting changes with this PR just to see how they do not work out once implemented. I think the best way forward is to go for the simplest and smallest fix possible to resolve this issue. If we want to actually fix things, we need to refactor code and that should come after fixing this bug.

I totally feel you on this.

Ok, I'm going to do the following: 

- As you suggest, I add an appropriate `drop` call in `main.rs`, with a comment saying why we need it to fix this bug. I already know where this needs to go.
- I undo all the rest of the changes in this PR.
- I'll make a note of the things we came across here in #3085 so we can integrate that into the cleanup plan.

> We already have a channel which we use in event.rs to write info to the pty, I wonder if we could use that same channel to send resize messages through. Then the event_loop.rs would be the sole owner of the pty and we could just read new messages from the channel and directly call pty.resize(xxx). That way we do not need an Arc anymore and all our problems would be gone.

This is an interesting thought. Would I be right in understanding this would be a change for all platforms, not just Windows? (Even if it helps solve our Windows issues.)

Reply:
> This is an interesting thought. Would I be right in understanding this would be a change for all platforms, not just Windows? (Even if it helps solve our Windows issues.)

Yes. On Linux the resize handle is just an integer, so it doesn't really matter what we do here.

We'd just change it on all platforms, since the code change that is required for Windows should automatically just work on Linux too.

Assuming that the order of the struct is in correct drop order, that should fix the problem without ever doing any `ConPTY` specific changes (other than the drop order thing). We'd basically just reduce complexity by getting rid of pointless multithreading support.